@uppy/react: deprecate `useUppy` (#4223)
transloadit/uppy: website/src/docs/react.md

npm install @uppy/react
```

## Usage

`@uppy/react` exposes component wrappers for `Dashboard`, `DragDrop`, and all other UI elements.
The components can be used with either [React][] or API-compatible alternatives such as [Preact][].

A couple things to keep in mind when using Uppy with React:

* Instead of adding a UI plugin to an Uppy instance with `.use()`, the Uppy instance can be passed into components as an `uppy` prop.
* All other props are passed as options to the plugin.
* The Uppy instance should **not** live inside the component but outside of it.

> * You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Reviewer: ```suggestion
* You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.
* An Uppy instance cannot be used by multiple components concurrently - be sure not to render more than one instance of a React component that uses an Uppy instance
```

npm install @uppy/react
```

## Usage

`@uppy/react` exposes component wrappers for `Dashboard`, `DragDrop`, and all other UI elements.
The components can be used with either [React][] or API-compatible alternatives such as [Preact][].

A couple things to keep in mind when using Uppy with React:

* Instead of adding a UI plugin to an Uppy instance with `.use()`, the Uppy instance can be passed into components as an `uppy` prop.
* All other props are passed as options to the plugin.
* The Uppy instance should **not** live inside the component but outside of it.
* You have to pass the IDs of your `use`d plugins to the `plugins` array prop so Dashboard knows it needs to render them.

> * An Uppy instance can’t be used by multiple components. Make sure you are using a unique Uppy instance per component.

Author: ```suggestion
* An Uppy instance can’t be used by more than one component. Make sure you are using a unique Uppy instance per component.
```

npm install @uppy/react
```

## Usage

The components can be used with either [React][] or API-compatible alternatives such as [Preact][].

Instead of adding a UI plugin to an Uppy instance with `.use()`, the Uppy instance can be passed into components as an `uppy` prop.
All other props are passed as options to the plugin.

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Author: I don't think this reasoning is required. 

- It is possible to make it work in React Strict Mode, we just chose not to do it.
- If a new user comes along, they never had to call `close` in general, so writing that you don't need to call it is unnecessary

I'll write a warning about using uppy in the render loop though 👍

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Reviewer: Did we really find a way that also lets us close uppy on unmount though? I tried to implement such a hook and document it but I failed because in strict mode we don’t know when a component is unmounted. Unmount is not a concept anymore with strict mode. If there is a way of doing it I think we should give an example in our docs because it is really *not* trivial!

as for close, maybe we dhould instead write it under uppy’s close-documentation (that people normally don’t need to call that function and it will make the uppy unusable after it’s closed)?

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Author: I don't think we want to provide an example because it would only encourage what we want to discourage? Regarding close, I think we might be confused because we worked a long time hiding it in the hook, but note that people never had this confusion. It was hidden in the hook and we never encouraged in the docs to always call it, so I don't see why we now need to warn to not do it.

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Reviewer: I still think it makes sense to inform users that they should use a single uppy instance, so they don’t waste times trying to inplement hooks and Uppy as React state themselves. 

I think we should document close somewhere. As in what it does and when to use it, but we dont need to document it here under «react»

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Author: > I still think it makes sense to inform users that they should use a single uppy instance, so they don’t waste times trying to inplement hooks and Uppy as React state themselves.

But we have that:

> * The Uppy instance should **not** live inside the component but outside of it.

What else do we need?

> You have to pass the IDs of your `use`d plugins to the `plugins` array props so Dashboard knows it needs to render them.

Reviewer: i missed that one sry!

All other props are passed as options to the plugin.

```js
import React, { useEffect } from 'react'
import Uppy from '@uppy/core'
import { Dashboard } from '@uppy/react'

const uppy = new Uppy()

function Component () {
  useEffect(() => {
    return () => {
      uppy.close({ reason: 'unmount' })
    }
  }, [])

  return <Dashboard uppy={uppy} />
}
```

### Dynamic options

```js
import React, { useState, useEffect } from 'react'
import Uppy from '@uppy/core'

> import { Dashboard } from '@uppy/react'

Reviewer 2: Do we want to highlight `import Dashboard from '@uppy/react/lib/Dashboard.js'`? We probably want to introduce an exports map to sugar code it, but importing the individual modules makes the most sense otherwise you have to fight with peer dependencies.

Author: I was hoping to do a separate PR for the exports map next to this so we don't have to document that workaround

const uppy = new Uppy()

function Component ({ restrictions }) {
  useEffect(() => {
    // Change @uppy/core options
    uppy.setOptions({ restrictions })

    // Or change some plugin dynamically
    uppy.getPlugin('SomePlugin').setOptions({ /* options */ })

    return () => {
      uppy.close({ reason: 'unmount' })
    }
  }, [restrictions])

  return <Dashboard uppy={uppy} />

> }

Reviewer 3: The Uppy instance here is shared between all instances of `Component` currently on the page. If I do:

```javascript
<Component restrictions={{allowedFileTypes: ['image/jpg']}} />

<Component restrictions={{allowedFileTypes: ['application/pdf']}} />
```

Won't they interfere with each other?

Author: Good point, in that case you want to do this:

```js
function Component ({ restrictions }) {
  const [uppy] = useState(() => new Uppy())

  useEffect(() => {
    uppy.setOptions({ restrictions })

    return () => {
      uppy.close({ reason: 'unmount' })
    }
  }, [restrictions])

  return <Dashboard uppy={uppy} />
}
```

This should also be added to the docs indeed. Thanks for pointing it out.
